### PR TITLE
1 more doc revamp

### PR DESCRIPTION
**Pull request by msaroufim**
Big revamp before PTC, will probably do a few more. but repo has changed quite a bit recently so this reflects the work people are actually using

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,138 +2,140 @@
...
-#### Without intrusive code changes
+### Post Training Quantization
 
-Quantizing your models is a 1 liner that should work on any model with an `nn.Linear` including your favorite HuggingFace model. You can find a more comprehensive usage instructions [here](torchao/quantization/) and a HuggingFace inference example [here](scripts/hf_eval.py)
+Quantizing your models is a 1 liner that should work on any model with an `nn.Linear` including your favorite HuggingFace model. You can find a more comprehensive usage instructions [here](torchao/quantization/), sparsity [here](/torchao/_models/sam/README.md) and a HuggingFace inference example [here](scripts/hf_eval.py)
+
+For inference we have the option of
+1. Quantize only the weights: works best for memory bound models
+2. Quantize the weights and activations: works best for compute bound models
+2. Quantize the activations and parameters and sparsify the parameters
```
Comment (supriyar) [inline README.md:27]: what do you mean by parameters here?

File: README.md [hunk 2]
```diff
@@ -2,138 +2,140 @@
...
+torchao just works with `torch.compile()` and `FSDP2` over most PyTorch models on Huggingface out of the box.
 
-### Inference
+## Inference
 
-#### Without intrusive code changes
+### Post Training Quantization
 
-Quantizing your models is a 1 liner that should work on any model with an `nn.Linear` including your favorite HuggingFace model. You can find a more comprehensive usage instructions [here](torchao/quantization/) and a HuggingFace inference example [here](scripts/hf_eval.py)
+Quantizing your models is a 1 liner that should work on any model with an `nn.Linear` including your favorite HuggingFace model. You can find a more comprehensive usage instructions [here](torchao/quantization/), sparsity [here](/torchao/_models/sam/README.md) and a HuggingFace inference example [here](scripts/hf_eval.py)
```
Comment (supriyar) [inline README.md:22]: why sparsity instructions in the quantizations section?

File: README.md [hunk 3]
```diff
@@ -2,138 +2,140 @@
...
-|             | int4wo-64-GPTQ     | 12.489              |  199.86       |  746.66                 |  4.50            |  3.74           |
-|             | autoquant          | 12.204              |  159.22       | 1069.87                 |  8.91            |  6.72           |
-| Llama-3-8B  | Base (bfloat16)    | N/A                 |   94.97       | 1425.55                 | 16.43            | 15.01           |
-|             | int8dq             | N/A                 |    8.44       |   63.45                 |  8.98            |  7.52           |
-|             | int8wo             | N/A                 |  139.76       | 1051.02                 | 10.42            |  7.52           |
-|             | int4wo-64          | N/A                 |  179.44       |  757.60                 |  6.62            |  4.22           |
-|             | autoquant          | N/A                 |  137.71       | 1037.74                 | 11.08            |  7.54           |
+We also a developer facing API so you can implement your own quantization algorithms so please use the excellent [HQQ](https://github.com/pytorch/ao/tree/main/torchao/prototype/hqq) algorithm as a motivating example.
```
Comment (supriyar) [inline README.md:42]: typo: we also provide

File: README.md [hunk 4]
```diff
@@ -2,138 +2,140 @@
...
-from torchao.quantization.quant_api import quantize_, int4_weight_only
-quantize_(m, int4_weight_only())
+from torchao.quantization.quant_api import quantize_, int8_dynamic_activation_int4_weight, int8_dynamic_activation_int8_weight, int8_dynamic_activation_int8_semi_sparse_weight, int4_weight_only, int8_weight_only
+quantize_(m, int4_weight_only()) 
 ```
 
-Benchmarks are run on a machine with a single A100 GPU using the script in `_models/llama` which generates text in a latency-optimized way (batchsize=1)
+For gpt-fast `int4_weight_only()` is the best option at bs=1 as it **2x the tok/s and reduces the VRAM requirements by about 65%** over a torch.compiled baseline 
+
+If you're unsure which option to use, you can also run autoquant which will automatically profile layers for you and skip quantizing layers where overhead is too large.
```
Comment (supriyar) [inline README.md:36]: link to autoquant readme?

File: README.md [hunk 5]
```diff
@@ -2,138 +2,140 @@
...
 
-note: Int8 dynamic quantization works best on compute bound as opposed to memory bound models. Some relatable examples might be [SAM](https://github.com/pytorch-labs/segment-anything-fast) which is compute bound vs Llama at batchsize=1 which is memory bound.
+### Quantization Aware Training
 
-For int4 we make heavy use of [tinygemm](https://github.com/pytorch/ao/blob/cb3bd8c674f2123af232a0231b5e38ddafa756a8/torchao/dtypes/aqt.py#L526) of `torch.ops.aten._weight_int4pack_mm` to bitpack into a layout optimized for tensor cores
+Quantizing a model in this way is called Post Training Quantization and it can sometimes give you a fast, small but innacurate model. To fix this we recommend you try out Quantization Aware Training and we shared a recipe in collaboration with the torchtune team [here](https://pytorch.org/blog/quantization-aware-training/) where we recovered up to **96% of the accuracy degradation on hellaswag and 68% of the perplexity degradation on wikitext** for Llama3 compared to post-training quantization (PTQ).
```
Comment (supriyar) [inline README.md:46]: suggest rephrasing this to something like 
"Using post-training quantization can result in a fast and compact model, but may also lead to accuracy degradation. To overcome this limitation, we recommend exploring Quantization Aware Training (QAT). In collaboration with Torchtune, we've developed a QAT recipe that demonstrates significant accuracy improvements over traditional PTQ, recovering..."

File: README.md [hunk 6]
```diff
@@ -2,138 +2,140 @@
...
-|            | int8 dynamic quant (attn)<br>int8 dynamic quant + 2:4 sparsity (mlp lin1)<br>2:4 sparsity (mlp lin2) | 26.46 | 14865        | 0.5668              | **1.16x**        | **97.54%**        |
+### Float8
 
-To reproduce our benchmarks please follow these [instructions](/torchao/_models/sam/README.md).
-
-#### With intrusive code changes
+[torchao.float8](torchao/float8) implements training recipes with the scaled float8 dtypes, as laid out in https://arxiv.org/abs/2209.05433.
 
-In some cases we rewrote popular GenAI models to be significantly faster in native PyTorch as in no C++/CUDA to achieve at the time SOTA inference performance. These involve more intrusive code changes.
+With ``torch.compile`` on, initial results show throughput speedups of up to **1.2x on small scale (8 GPUs) LLaMa pretraining jobs**.
```
Comment (supriyar) [inline README.md:71]: any benchmark numbers for this?

File: README.md [hunk 7]
```diff
@@ -2,138 +2,140 @@
...
 
-* [MX](torchao/prototype/mx_formats) implementing training and inference support with tensors using the [OCP MX spec](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf) data types, which can be described as groupwise scaled float8/float6/float4/int8, with the scales being constrained to powers of two. This work is prototype as the hardware support is not available yet.
-* [nf4](torchao/dtypes/nf4tensor.py) which was used to [implement QLoRA](https://github.com/pytorch/torchtune/blob/main/docs/source/tutorials/qlora_finetune.rst) one of the most popular finetuning algorithms without writing custom Triton or CUDA code. Accessible talk [here](https://x.com/HamelHusain/status/1800315287574847701)
-* [fp6](torchao/prototype/quant_llm/) for 2x faster inference over fp16 with an easy to use API `quantize_(model, fp6_llm_weight_only())`
+We also have support for s[ingle GPU CPU offloading](https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim#optimizer-cpu-offload) where both the gradients (same size as parameters) and the optimizers will be efficiently sent to the CPU. This alone can **reduce your VRAM requirements by 60%**
```
Comment (supriyar) [inline README.md:102]: nit: s is outside the hyperlink

**Review by supriyar — APPROVED** [review]
left a few comments to address. Other than that lgtm!

File: README.md (hunk 1 above)
Comment (msaroufim) [reply README.md:27]: I mean the weights. Hmm I'll just weights then might be clearer
